flows/contact.go

```diff
@@ -401,13 +402,21 @@ func (c *Contact) UpdatePreferredChannel(channel *Channel) bool {
 		for _, urn := range c.urns {
 			// tel URNs can be re-assigned, other URN schemes are considered channel specific
 			if urn.URN().Scheme() == urns.TelScheme && channel.SupportsScheme(urns.TelScheme) {
-				urn.SetChannel(channel)
+				if channel.HasRole(assets.ChannelRoleSend) {
+					urn.SetChannel(channel)
+				} else {
+					errs = append(errs, errors.Errorf("unable to set channel on urn '%s', missing the send role on channel '%s'", urn.URN().String(), channel.UUID()))
```

Comment:
This seems more complicated than it needs to be. returning a separate error for every URN

Just have the action generate an error event instead of a modifier if the channel doesn't have a send role.

This method could start by checking if the channel has the send role once.. but that's not so important since it'll never be called with such a channel.

Reply:
so I can make that one error instead of one error per URN,
but I don't understand how to generate the error event instead of a modifier?

Reply:
You just need something like this in the action:

```
	if a.Channel != nil {
		channel = run.Session().Assets().Channels().Get(a.Channel.UUID)
		if channel == nil {
			logEvent(events.NewDependencyError(a.Channel))
			return nil
		}

		if !channel.HasRole(assets.ChannelRoleSend) {
			logEvent(events.NewErrorf("can't set channel that can't send as the preferred channel"))
			return nil
		}
	}
```

Reply:
never mind I get that now